### PR TITLE
Deals

### PR DESCRIPTION
**Pull request by evfinkn**
Checkout page showing discounts (Practical Plastic Car has a 50% off discount):

Screenshot 2023-12-09 at 16-30-07 Selt2023FinalProjectTeam001

### CODE REVIEW

File: app/models/promotion.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  def started?
+    starts_on < Time.now
+  end
+
+  def ended?
+    ends_on < Time.now
+  end
+
+  def active?
+    started? && !ended?
```
Comment (blhohmann) [inline app/models/promotion.rb:36]: Do you have this working with price_alerts, rn price alerts only checks when a price is manually updated (see here https://github.com/uiowahjmjohnsonselt2023/selt2023-final-project-team-001/blob/0d139671700150c07a339b35543722b58721bd0f/app/controllers/products_controller.rb#L98-L109)

If this doesn't work with price alerts, maybe you could throw something like that code here, where if the deal is active and the price is lower than any price alert thresholds, send the email and delete the price alert?

**Review by blhohmann — CHANGES_REQUESTED** [review]
Haven't looked over everything, just a comment about how this should interact with my merged PR of price_alerts

File: app/models/promotion.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  validates :ends_on, comparison: {greater_than: :starts_on}
+  # Minimum number of items required to qualify for the promotion
+  validates :min_quantity, numericality: {only_integer: true}
+  # Maximum number of items that can be discounted by the promotion. If non-positive,
+  # there is no limit.
+  validates :max_quantity, numericality: {only_integer: true}
+  validates :max_quantity,
+    comparison: {greater_than_or_equal_to: :min_quantity},
+    unless: -> { max_quantity <= 0 }
+
```
Comment (blhohmann) [inline app/models/promotion.rb:29]:
```suggestion
  validates :product_ids, presence: true
```
This makes it so you can't create a promotion without selecting a product to include.

File: app/views/products/show.html.erb [hunk 3]
```diff
@@ -6,7 +6,11 @@
```
Comment (blhohmann) [inline app/views/products/show.html.erb:6]: We should probably have something here to show a deal exists on the product, before we get to the checkout. I'll play around and see if I can get a suggestion.

**Review by blhohmann — CHANGES_REQUESTED** [review]
Other thing is we will need to add this to the user dropdown and rebase. Looks great!

File: app/models/promotion.rb (hunk 2 above)
Comment (evfinkn) [reply app/models/promotion.rb:29]: Actually this is intentional. Promotions without products selected will apply to all of the seller's products. I'll add a note on the form 🫡

**Review by blhohmann — APPROVED** [review]
Great changes, good work looks cool!

**Review by powersaudrey25 — APPROVED** [review]

**Pull request merged**